neuraxle/steps/flow.py

```diff
@@ -450,6 +450,27 @@ def _make_all_steps_optional(self):
 
         self._refresh_steps()
 
+    def prune(self) -> HyperparameterSamples:
+        """
+        Prune self so as to keep only the chosen step as set in the choice hyperparameter.
+        """
+        # first, get chosen step name. it looks like this:
+        chosen_step_name: str = self.get_hyperparams()["choice"]
```

Comment:
Is this the good hyperparam? That was from memory. I believe it is not. If my guess was right in my quick code, that is a miracle.

Reply:
that's true. self.get_hyperparams() does't not return the best hyperparams. Then i think `prune` method should not belong to ChooseOneStep class, It has to be under AutoML.prune because, that's where we get the information of the  best_model(the model which has the highest accuracy). So the idea goes like this
```
auto_ml = auto_ml.fit(X_train, y_train) # once model is fitted
auto_ml.apply("prune") #so under auto_ml we have the access to get the best model and trim all the remaining models.
best_pipeline = auto_ml.get_best_model() # only loads the best model and remaining all models are removed in the previous  step.
```
Do you agree with this @guillaume-chevalier ?

Reply:
@Rohith295, 

I have checked the code and we should replace `self.get_hyperparams()["choice"]` by `self.get_hyperparams()[self.CHOICE_HYPERPARAM]` and then the code should be ok. For instance you may want to read [this method](https://github.com/Neuraxio/Neuraxle/blob/e9bc7447649653d759e015827ba4c76f1c65b9b6/neuraxle/steps/flow.py#L424). 

It is not important that the chosen step be the best step from the point of view of the implementation of the `ChooseOneStepOf` step. The important thing is that this step does in fact only uses the step specified in the `self.CHOICE_HYPERPARAM` hyperparam as seen in the method linked in the paragraph just above. 

For what regards the AutoML, I don't think we should add the parameter for the simple reason that the AutoML class is already too complicated and already have too many hyperparameters. It is already subject to change as per the code imagined in the issue #504 to reduce the size of the class by splitting it into more different classes. 

I have an idea to solve the problem of the pruning: we could add an argument to the `ChooseOneOrManyStepsOf` class and the `ChooseOneStepOf` class in their `__init__` method, that would be like `ChooseOneStepOf(steps=[...], auto_prune=True)`. If auto_prune is set to true, then the `setup` method would call `self.prune()`. 

It is important to note that when the `setup` method is called, the hyperparameters are already set. Because the setup method is called at the moment of the first `fit` or the first `transform` of the pipeline, as seen in the [Machine Learning Objects Lifecycle Chart](https://www.neuraxle.org/stable/step_saving_and_lifecycle.html): 
![](https://www.neuraxle.org/stable/_images/neuraxle_machine_learning_lifecycle1.png)

Reply:
Also, to code the `ChooseOneOrManyStepsOf` class' pruning, it think that is should call Optional.prune on its children. See here: https://github.com/Neuraxio/Neuraxle/blob/e9bc7447649653d759e015827ba4c76f1c65b9b6/neuraxle/steps/flow.py#L519